### ISSUE TITLE
RoPE-Mixed implementation

### ISSUE THREAD

**Issue by orrzohar** [naver-ai/rope-vit#3]
Hi,

the implementation of RoPE-Mixed:
<img width="641" alt="image" src="https://github.com/user-attachments/assets/dad45700-85bc-40d8-b4d4-07bf486fc6fc">

seems to contradict your implementation here:

https://github.com/naver-ai/rope-vit/blob/c6aa201ee795daa4f841e2f9585164bb23a0b819/models/vit_rope.py#L52-L62

specifically, shouldn't you be applying the x and y frequencies on all embeddings, not the x on the first half and the y on the last half?

Best,
Orr

**Comment by bhheo** [comment]
Hi Orr

You can find the RoPE-Mixed implementation here:

https://github.com/naver-ai/rope-vit/blob/c6aa201ee795daa4f841e2f9585164bb23a0b819/models/vit_rope.py#L39-L50

The lines you linked are code for RoPE-Axial, i.e., Eq (12) in the paper

Best
Byeongho heo

**Comment by orrzohar** [comment]
Hi Byeongho heo,

Thank you for clarifying!
however, here is also appears that you are only applying the `x` position_ids on the first D//2 dimensions and the `y` positional_ids on the last D//2 dimension -- but in the equation, it looks like you are just adding two frequencies to all position_ids (one for x and the other for y). 


shouldn't the implementation be: 
```
 def compute_mixed_cis(freqs: torch.Tensor, t_x: torch.Tensor, t_y: torch.Tensor, num_heads: int): 
     N = t_x.shape[0] 
     depth = freqs.shape[1] 
     # No float 16 for this range 
     with torch.cuda.amp.autocast(enabled=False): 
         freqs_x = (t_x.unsqueeze(-1) @ freqs[0].unsqueeze(-2)).view(depth, N, num_heads, -1).permute(0, 2, 1, 3) 
         freqs_y = (t_y.unsqueeze(-1) @ freqs[1].unsqueeze(-2)).view(depth, N, num_heads, -1).permute(0, 2, 1, 3) 
         freqs_cis=torch.polar(torch.ones_like(freqs_x), freqs_x + freqs_y).unsqueeze(0)
                      
     return torch.cat(freqs_cis, dim=-1) 
    
```
Best,
Orr

**Comment by bhheo** [comment]
Oh, sorry for the confusion.

I think you might be confused because:
```python
freqs_cis = [] 
freqs_cis.append(torch.polar(torch.ones_like(freqs_x), freqs_x + freqs_y)) 
return torch.cat(freqs_cis, dim=-1) 
```

`freqs_cis = []` is a dummy code.
I used it when I was designing RoPE-Mixed. I was trying frequency sharing over different heads.
Anyway, it means nothing now, but I just forgot to delete it.

It is the same as 
```python
return torch.polar(torch.ones_like(freqs_x), freqs_x + freqs_y)
```
because there is only one element in `freqs_cis` list.
It is also the same as your code except `.unsqueeze(0)`

Please note that it is not for D//2 dim separation.

Best
Heo

**Comment by orrzohar** [comment]
Hi Heo,

OK great, last question (thank you for being so responsive!) -- do you happen to know if this edit is compatible with flash attention? 

Best,
Orr

**Comment by bhheo** [comment]
I know nothing about compatibility with flash-attention.

If the complex number qk is compatible with flash-attention, then ours might also be compatible.
Because our implementation follows' code-llama's RoPE, it might help you to find `code-llama's compatibility.

But I'm not sure about it. I don't have enough knowledge to answer it.
I hope you got a good result on your research or programming.

Best
Heo

**Comment by orrzohar** [comment]
Hi Heo,

Thank you, and best of luck on your future work! And great work v. RoPE-Mixed, I think this style of implementation makes the most sense. 

Best,
Orr
